File: Алгоритм Форда-Фалкерсона/ford_fulkerson.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
#include <cstring>
using namespace std;

class FordFulkerson {
private:
    int n; // Количество вершин
    vector<vector<int>> capacity; // Матрица пропускных способностей
    vector<vector<int>> adj;      // Список смежности для графа

public:
    FordFulkerson(int n) : n(n), capacity(n, vector<int>(n, 0)), adj(n) {}

    void add_edge(int u, int v, int cap) {
        adj[u].push_back(v);
        adj[v].push_back(u);
        capacity[u][v] += cap; // Поддержка нескольких рёбер между вершинами
    }

    int bfs(int source, int sink, vector<int>& parent) {
        fill(parent.begin(), parent.end(), -1);
        parent[source] = source;
        queue<pair<int, int>> q;
        q.push({source, INT_MAX});

        while (!q.empty()) {
            int current = q.front().first;
            int flow = q.front().second;
            q.pop();

            for (int next : adj[current]) {
                if (parent[next] == -1 && capacity[current][next] > 0) {
                    parent[next] = current;
                    int new_flow = min(flow, capacity[current][next]);
                    if (next == sink) {
                        return new_flow;
                    }
                    q.push({next, new_flow});
                }
            }
        }
        return 0;
    }

    int max_flow(int source, int sink) {
        int flow = 0;
        vector<int> parent(n);
        int new_flow;

        while (new_flow = bfs(source, sink, parent)) {
            flow += new_flow;
            int current = sink;
            while (current != source) {
                int prev = parent[current];
                capacity[prev][current] -= new_flow;
                capacity[current][prev] += new_flow;
                current = prev;
            }
        }
        return flow;
    }
};

#ifndef TESTING // Условная компиляция для отключения main()
int main() {
    int n, m;
    cin >> n >> m; // Чтение количества вершин и рёбер
    FordFulkerson ff(n);

    for (int i = 0; i < m; ++i) {
        int u, v, capacity;
        cin >> u >> v >> capacity;
        ff.add_edge(u, v, capacity);
    }

    int source, sink;
    cin >> source >> sink; // Чтение истока и стока

    cout << "Max Flow: " << ff.max_flow(source, sink) << endl;

    return 0;
}
#endif

File: ford_fulkerson/ford_fulkerson.py
from collections import defaultdict, deque

class FordFulkerson:
    def __init__(self, n):
        self.n = n  # Количество вершин
        self.graph = defaultdict(list)  # Список смежности для графа
        self.capacity = {}  # Пропускные способности рёбер

    def add_edge(self, u, v, capacity):
        """Добавляет ребро с пропускной способностью."""
        self.graph[u].append(v)
        self.graph[v].append(u)
        self.capacity[(u, v)] = capacity
        self.capacity[(v, u)] = 0  # Обратное ребро с нулевой пропускной способностью

    def bfs(self, source, sink, parent):
        """Поиск пути увеличения потока с использованием BFS."""
        visited = [False] * self.n
        queue = deque([source])
        visited[source] = True
        while queue:
            current = queue.popleft()
            for neighbor in self.graph[current]:
                # Ищем ребро с ненулевой остаточной пропускной способностью
                if not visited[neighbor] and self.capacity[(current, neighbor)] > 0:
                    parent[neighbor] = current
                    if neighbor == sink:
                        return True
                    visited[neighbor] = True
                    queue.append(neighbor)
        return False

    def max_flow(self, source, sink):
        """Вычисляет максимальный поток."""
        flow = 0
        parent = [-1] * self.n
        while self.bfs(source, sink, parent):
            # Найти минимальную пропускную способность вдоль пути увеличения
            path_flow = float('inf')
            current = sink
            while current != source:
                prev = parent[current]
                path_flow = min(path_flow, self.capacity[(prev, current)])
                current = prev

            # Обновить остаточные пропускные способности рёбер
            current = sink
            while current != source:
                prev = parent[current]
                self.capacity[(prev, current)] -= path_flow
                self.capacity[(current, prev)] += path_flow
                current = prev

            flow += path_flow
        return flow

def main():
    """Пример использования алгоритма Форда-Фалкерсона."""
    n, m = map(int, input().split())  # Чтение количества вершин и рёбер
    ff = FordFulkerson(n)
    for _ in range(m):
        u, v, capacity = map(int, input().split())
        ff.add_edge(u, v, capacity)
    source, sink = map(int, input().split())  # Исток и сток
    print("Максимальный поток:", ff.max_flow(source, sink))

if __name__ == "__main__":
    main()
